theano/gpuarray/sort.py

```diff
@@ -0,0 +1,314 @@
+        reordered_axes = list(range(ndim))
+        axis = self.axis % ndim
+        del(reordered_axes[axis])
+        reordered_axes = [axis] + reordered_axes
+        dims = ''.join('(void*)(dims+%d), ' % i for i in reordered_axes[1:])
+        prep_output = ''
+        if self.return_values:
+            def_dvstrides = 'const ssize_t *dvstrides = PyGpuArray_STRIDES(%s)' % yv
+            params_dv = '(void*)(%s->ga.data),\n' % yv
+            params_dv += ''.join('(void*)(dvstrides+%d), ' % i for i in reordered_axes)
+            prep_output += '''
+    if (0 != theano_prep_output(
+        &%(yv)s, %(ndim)d, odims,
+        %(inp_dtc)s, GA_C_ORDER, %(ctx)s)) {
+        %(fail)s;
+    }\n''' % locals()
+        else:
+            def_dvstrides = params_dv = ''
+
+        if self.return_indices:
+            def_distrides = 'const ssize_t *distrides = PyGpuArray_STRIDES(%s)' % yi
+            params_di = '(void*)(%s->ga.data),\n' % yi
+            params_di += ''.join('(void*)(distrides+%d), ' % i for i in reordered_axes)
+            prep_output += '''
+    if (0 != theano_prep_output(
+        &%(yi)s, %(ndim)d, odims,
+        %(out_dtc)s, GA_C_ORDER, %(ctx)s)) {
+        %(fail)s;
+    }\n''' % locals()
+        else:
+            def_distrides = params_di = ''
+        sstrides = ', '.join('(void*)(sstrides+%d)' % i for i in reordered_axes)
+        code = '''
+{
+    const ssize_t k_ = ((%(k_dtype)s*)(PyArray_DATA(%(k)s)))[0];
+    const size_t *dims = PyGpuArray_DIMS(%(x)s);
+    size_t odims[%(ndim)d];
+    for (int i=0; i<%(ndim)d; i++)
+        odims[i] = dims[i];
+
+    odims[%(axis)d] = k_>=0 ? k_ : -k_;
+
+    if (0 == odims[%(axis)d]) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "topk: kth must not be zero");
+        %(fail)s;
+    } else if (dims[%(axis)d] < odims[%(axis)d]) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "topk: kth cannot be larger than the size of specified axis %(axis)d");
+        %(fail)s;
+    }
+    %(prep_output)s
+
+    size_t blk[6];
+    size_t *grd = blk+3;
+    blk[0] = blk[1] = blk[2] = 1;
+    grd[0] = grd[1] = grd[2] = 1;
+    for (int i=0; i<%(ndim)d; ++i) {
+        if (i!=%(axis)d)
+            grd[0] *= dims[i];
+        else
+            blk[0] = dims[i];
+    }
+    // round up to multiples of warp size
+    blk[0] = ((blk[0] + %(WARP_SIZE)d - 1) / %(WARP_SIZE)d) * %(WARP_SIZE)d;
+
+    %(def_dvstrides)s;
+    %(def_distrides)s;
+    const ssize_t *sstrides = PyGpuArray_STRIDES(%(x)s);
+    void* args[] = {
+        %(dims)s
+        %(params_dv)s
+        %(params_di)s
+        (void*)(&k_),
+        (void*)(%(x)s->ga.data),
+        %(sstrides)s,
+        (void*)(dims+%(axis)d),
+    };
+
+    int err;
+    if (dims[%(axis)d] > PY_SSIZE_T_MAX) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "topk: array size on specified axis is too large, should be less than PY_SSIZE_T_MAX.");
+        %(fail)s;
```

Comment:
This situation is not an error for GpuArrays (although I really doubt you will encounter any array that big).